docs/rfc/block-registration.md

```diff
@@ -0,0 +1,383 @@
+
+A demonstration for how block registration could be made filterable in PHP was explored in [#5802](https://github.com/WordPress/gutenberg/pull/5802). The purpose here was to explore how plugins could have better control over the registration.
+
+Another exploration in [#5652](https://github.com/WordPress/gutenberg/pull/5652) considered using JSON as a file format to share block type definitions between JavaScript and PHP. 
+
+### Conclusions
+
+* The current approaches to client-side block type registration cannot support the proposed requirement to have all block types known outside the browser context.
+* Using a statically-defined, JSON-formatted block type definition enables easy integration in both JavaScript and PHP runtimes.
+* Registering a block type in PHP would allow for attribute default values to be assigned as dynamically generated from some external state (e.g. a database value, or localized string).
+* By default, JSON does not support localization or dynamic values.
+* On the server, a block type `icon` property can only be assigned as a string and thus cannot support SVGs and component-based icons.
+
+---
+
+# Block Type Registration RFC
+
+## Introduction
+
+Blocks are the fundamental elements of the editor. They are the primary way in which plugins and themes can register their own functionality and extend the capabilities of the editor.
+
+## Registering a block type
+
+To register a new block type, start by creating a `block.json` file. This file:
+
+* Gives a name to the block type.
+* Defines some important metadata about the registered block type (title, category, icon, description, keywords).
+* Defines the attributes of the block type.
+* Links to the editor implementation of your block type, the save function and any other context-aware property.
+
+**Example:**
+
+
+```json
+{
+	"name": "my-plugin/notice",
+	"title": "Notice",
+	"category": "common",
+	"icon": "star",
+	"description": "Shows warning, error or success notices  ...",
+	"keywords": [ "alert", "message" ],
+	"attributes": {
+		"message": {
+			"type": "string",
+			"source": "html",
+			"selector": ".meessage"
+		}
+	},
+	"edit": "blocks/notice-edit.js",
```

Comment:
What are the long term expectations of this separation? Is it anticipated that "edit" could be plural based on environments (like mobile native)? Is it useful to specify these (implementation?) details on the definition file?

Reply:
It's more designed this way to be able to load the asset containing `edit` definition only when the block is being inserted into the editor.

Reply:
I can see a potential "edit:native". In a similar way, I see `render_callback:drupal`.